contributors/design-proposals/hugepages.md

```diff
@@ -0,0 +1,213 @@
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: example
+spec:
+  containers:
+...
+    resources:
+      requests:
+	    hugepages: "10"
+      limits:
+	    hugepages: "10"
+  nodeSelector:
+    kubernetes.io/huge-page-size: "2MB"
+```
+
+Huge pages can not be overcommitted on a node.
+
+While a system may support multiple huge pages sizes, it is assumed that nodes
+configured with huge pages will only use one huge page size, namely the default
+page size in `cat /proc/meminfo | grep Hugepagesize`.  In Linux, this is 2MB
+unless overridden by `default_hugepagesz=1g` in the kernel boot parameters.
+
+The huge page size for the node will be reported by the kubelet as a label
+`alpha.kubernetes.io/huge-page-size` on the node resource.  This is done
+because there are a variety of huge page sizes across different hardware
+architecture and making a new resource field for each size doesn't scale.  Pods
+can do a nodeSelector on this label to land on a system with a particular huge
+page size.  This is similiar to how the `beta.kubernetes.io/arch` label
+operates.
+
+## Limits and Quota
+
+LimitRange should be able to define minimum and maximum constraints for huge
+pages, and Quota should be able to count them.
+
+## Implementation
+
+### Phase 0: Design Agreement
+
+**Target 1.5**
+
+Get design approval
+
+### Phase 1: Add huge page support
+
+**Target 1.5+**
+
+Implement huge page support with pod-level cgroups to enforce per-pod huge page
+limits (not yet available).  Enforcing huge page limits with pod-level cgroups
+avoids, at least temporarily, the need for 1) `docker` to support the
+`hugetlb` cgroup controller directly and 2) adding huge pages to the
+Container Runtime Interface (CRI)
+
+pkg/api/types.go (and v1/types.go)
+
+```
+const (
+	// CPU, in cores. (500m = .5 cores)
+	ResourceCPU ResourceName = "cpu"
+	// Memory, in bytes. (500Gi = 500GiB = 500 * 1024 * 1024 * 1024)
+	ResourceMemory ResourceName = "memory"
+...
+	ResourceHugePages ResourceName = "hugepages"
+)
+```
+
+The kubelet will report the total/available huge pages statistics from cadvisor
+in the node status as the huge page capacity/available.
+
+Modifications needed `setNodeStatusMachineInfo` in `pkg/kubelet/kubelet_node_status.go`
+and `CapacityFromMachineInfo` in `pkg/kubelet/cadvisor/util.go`.
+
+The kubelet will also need to create the `alpha.kubernetes.io/huge-page-size`
+label for its node resource (if self registering).
+
+pkg/api/unversioned/well_known_labels.go
+
+```
+const (
+...
+	LabelArch = "beta.kubernetes.io/arch"
+	LabelHugePageSize = "alpha.kubernetes.io/huge-page-size"
+)
+```
+
+The scheduler will need to ensure any huge page request defined in the pod spec can be fulfilled by a candidate node.
+
+cAdvisor will need to be modified to return the number of available huge pages.
+This is already supported in [runc/libcontainer](../../vendor/github.com/opencontainers/runc/libcontainer/cgroups/utils.go)
+
+### Phase 2: Expose huge pages in CRI
```

Comment:
Can you explain why this is desirable rather than just sticking with the pod-level implementation as above? In the abstract you talked about it as a pod feature and this jump is unclear.